Add "Add email address" custom flow guide (#1197)
clerk/clerk-docs: docs/custom-flows/add-email.mdx

---
description: Learn how to use Clerk's API to build a custom flow for adding an email to a user's account.
---

# Build a custom flow for adding an email to a user's account

<Callout type="danger">
  This guide is for users who want to build a *custom* user interface using the Clerk API. To use a *prebuilt* UI, you should use Clerk's [Account Portal pages](/docs/account-portal/overview) or [prebuilt components](/docs/components/overview).
</Callout>

Users are able to add multiple email addresses to their account.

Every user has a [`User`](/docs/references/javascript/user/user) object that represents their account. The `User` object has a `emailAddresses` property that contains all the email addresses associated with the user.

> The following example demonstrates how to build a custom user interface that allows users to add an email address to their account. The example:

Author: It might be worth mentioning that this is a custom version of the `add phone number` function from the `<UserProfile />`

1. Uses the `prepareVerification()` method on the newly created `EmailAddress` object to send a verification code to the user.
1. Uses the `attemptVerification()` method on the `EmailAddress` object with the verification code provided by the user to verify the email address.

<Tabs items={["Next.js"]}>
  <Tab>
  ```jsx {{ filename: 'app/account/add-email/page.tsx' }}
  'use client';

  import * as React from 'react';
  import { useUser } from '@clerk/nextjs';
  import { EmailAddressResource } from '@clerk/types';

  export default function Page() {
    const { isLoaded, user } = useUser();
    const [email, setEmail] = React.useState('');
    const [code, setCode] = React.useState('');
    const [isVerifying, setIsVerifying] = React.useState(false);
    const [successful, setSuccessful] = React.useState(false);
    const [emailObj, setEmailObj] = React.useState<
      EmailAddressResource | undefined
    >();

    if (!isLoaded) return null;

    if (isLoaded && !user?.id) {
      return <p>You must be logged in to access this page</p>;
    }

    // Handle addition of the email address
    const handleSubmit = async (e: React.FormEvent) => {
      e.preventDefault();

      try {

>         // Add email address to user

Author: ```suggestion
        // Add an unverified email address to user
```

        const res = await user.createEmailAddress({ email });

>         // Reload user to get updated user info

Author: ```suggestion
        // Reload user to get updated user object
```

        await user.reload();

>         // Find the email address that was just added

Author: ```suggestion
        // Create a reference to the email address that was just added
```

        const emailAddress = user.emailAddresses.find((a) => a.id === res.id);

>         setEmailObj(emailAddress);

Reviewer: ```suggestion
        // Create a reference to the email address that was just added
        setEmailObj(emailAddress);
```

        // Send the user an email with the verification code
        emailAddress?.prepareVerification({ strategy: 'email_code' });

        // Set to true to display second form
        // and capture the OTP code
        setIsVerifying(true);
      } catch (err) {
        // See https://clerk.com/docs/custom-flows/error-handling
        // for more info on error handling
        console.error(JSON.stringify(err, null, 2));
      }
    };

    // Handle the submission of the verification form
    const verifyCode = async (e: React.FormEvent) => {
      e.preventDefault();
      try {

>         // Use the code the user provided to attempt verification

Author: ```suggestion
        // Verify that the code entered matches the code sent to the user
```
